File: mobile/android/core/auth/src/main/kotlin/com/slopssaloon/omen/core/auth/SupabaseAuthRepository.kt
package com.slopssaloon.omen.core.auth

import com.slopssaloon.omen.core.session.SecureSessionStore
import com.slopssaloon.omen.core.session.Session

/**
 * Live [AuthRepository] backed by Supabase GoTrue via a [GoTrueTransport]. Maps transport
 * results to opaque [AuthOutcome]s; the app never sees a raw provider error (M0c §8).
 *
 * `nowEpochSeconds` is injected so token-expiry math is deterministic in unit tests.
 * `store` supplies the refresh token for [refresh]; persistence remains the SessionManager's job.
 */
class SupabaseAuthRepository(
    private val transport: GoTrueTransport,
    private val store: SecureSessionStore,
    private val nowEpochSeconds: () -> Long,
) : AuthRepository {

    override suspend fun requestEmailOtp(email: String): AuthOutcome = when (val r = transport.requestEmailOtp(email)) {
        is TransportResult.Ok, is TransportResult.SessionTokens -> AuthOutcome.OtpSent
        is TransportResult.HttpError -> AuthOutcome.RetryableError(r.status.toRetryable())
        TransportResult.NetworkError -> AuthOutcome.RetryableError(RetryableCode.NETWORK)
        TransportResult.Malformed, is TransportResult.Challenge -> AuthOutcome.RetryableError(RetryableCode.UNKNOWN)
    }

    override suspend fun verifyEmailOtp(email: String, code: String): AuthOutcome =
        when (val r = transport.verifyEmailOtp(email, code)) {
            is TransportResult.SessionTokens -> AuthOutcome.Success(r.toSession())
            is TransportResult.HttpError -> if (r.status in 400..403) AuthOutcome.InvalidCode
                else AuthOutcome.RetryableError(r.status.toRetryable())
            TransportResult.NetworkError -> AuthOutcome.RetryableError(RetryableCode.NETWORK)
            TransportResult.Ok, TransportResult.Malformed, is TransportResult.Challenge ->
                AuthOutcome.RetryableError(RetryableCode.UNKNOWN)
        }

    override suspend fun signInWithGoogleIdToken(idToken: String, rawNonce: String): AuthOutcome =
        when (val r = transport.signInWithGoogleIdToken(idToken, rawNonce)) {
            is TransportResult.SessionTokens -> AuthOutcome.Success(r.toSession())
            is TransportResult.HttpError -> if (r.status in 400..403) AuthOutcome.Unsupported
                else AuthOutcome.RetryableError(r.status.toRetryable())
            TransportResult.NetworkError -> AuthOutcome.RetryableError(RetryableCode.NETWORK)
            TransportResult.Ok, TransportResult.Malformed, is TransportResult.Challenge ->
                AuthOutcome.RetryableError(RetryableCode.UNKNOWN)
        }

    override suspend fun refresh(): AuthOutcome {
        val refreshToken = store.load()?.refreshToken ?: return AuthOutcome.NeedsReauth
        return when (val r = transport.refresh(refreshToken)) {
            is TransportResult.SessionTokens -> AuthOutcome.Success(r.toSession())
            is TransportResult.HttpError -> AuthOutcome.NeedsReauth
            TransportResult.NetworkError -> AuthOutcome.RetryableError(RetryableCode.NETWORK)
            TransportResult.Ok, TransportResult.Malformed, is TransportResult.Challenge -> AuthOutcome.NeedsReauth
        }
    }

    override suspend fun exchangeOAuthCode(
        providerId: String,
        code: String,
        codeVerifier: String,
    ): AuthOutcome = when (val r = transport.exchangeOAuthCode(providerId, code, codeVerifier)) {
        is TransportResult.SessionTokens -> AuthOutcome.Success(r.toSession())
        // 400 on OAuth exchange means either the code is stale/reused (mismatch) or the provider
        // isn't configured; Supabase distinguishes these only in the response body, which by
        // §M0c-8 the transport must not expose. Treat 400/403 as a callback-mismatch class error;
        // 404 means the provider isn't wired.
        is TransportResult.HttpError -> when (r.status) {
            in 400..403 -> AuthOutcome.OAuthCallbackMismatch
            404 -> AuthOutcome.OAuthProviderNotConfigured
            else -> AuthOutcome.RetryableError(r.status.toRetryable())
        }
        TransportResult.NetworkError -> AuthOutcome.RetryableError(RetryableCode.NETWORK)
        TransportResult.Ok, TransportResult.Malformed, is TransportResult.Challenge ->
            AuthOutcome.RetryableError(RetryableCode.UNKNOWN)
    }

    override suspend fun startPasskeyChallenge(): PasskeyChallenge =
        when (val r = transport.startPasskeyChallenge()) {
            is TransportResult.Challenge -> PasskeyChallenge.Ok(r.challenge)
            is TransportResult.HttpError -> PasskeyChallenge.Failed(r.status.toRetryable())
            TransportResult.NetworkError -> PasskeyChallenge.Failed(RetryableCode.NETWORK)
            TransportResult.Ok, TransportResult.Malformed, is TransportResult.SessionTokens ->
                PasskeyChallenge.Failed(RetryableCode.UNKNOWN)
        }

    override suspend fun signInWithPasskey(assertion: PasskeyResult.Assertion): AuthOutcome =
        when (val r = transport.verifyPasskeyAssertion(assertion)) {
            is TransportResult.SessionTokens -> AuthOutcome.Success(r.toSession())
            is TransportResult.HttpError -> AuthOutcome.RetryableError(r.status.toRetryable())
            TransportResult.NetworkError -> AuthOutcome.RetryableError(RetryableCode.NETWORK)
            TransportResult.Ok, TransportResult.Malformed, is TransportResult.Challenge ->
                AuthOutcome.RetryableError(RetryableCode.UNKNOWN)
        }

    override suspend fun registerPasskey(credential: PasskeyResult.Assertion): AuthOutcome =
        when (val r = transport.registerPasskey(credential)) {
            // Registration returns 2xx-with-no-body on success; the app already has a session,
            // so we don't need new tokens here. Map to a no-op Success carrying the current one.
            is TransportResult.Ok -> store.load()?.let { AuthOutcome.Success(it) }
                ?: AuthOutcome.NeedsReauth
            is TransportResult.SessionTokens -> AuthOutcome.Success(r.toSession())
            is TransportResult.HttpError -> AuthOutcome.RetryableError(r.status.toRetryable())
            TransportResult.NetworkError -> AuthOutcome.RetryableError(RetryableCode.NETWORK)
            TransportResult.Malformed, is TransportResult.Challenge -> AuthOutcome.RetryableError(RetryableCode.UNKNOWN)
        }

    override suspend fun signOut() {
        // Best-effort remote sign-out is optional; local secure storage is cleared by SessionManager.
    }

    private fun TransportResult.SessionTokens.toSession() = Session(
        userId = userId,
        accessToken = accessToken,
        refreshToken = refreshToken,
        expiresAtEpochSeconds = nowEpochSeconds() + expiresInSeconds,
    )

    private fun Int.toRetryable(): RetryableCode = when {
        this == 408 || this == 504 -> RetryableCode.TIMEOUT
        this in 500..599 -> RetryableCode.SERVER
        else -> RetryableCode.UNKNOWN
    }
}

File: mobile/ios/OmenIOS/OmenIOS/Core/Auth/SupabaseAuthRepository.swift
import Foundation

/// Mirrors Android `core/auth/SupabaseAuthRepository.kt`'s status→outcome mapping rules exactly
/// (behavioral parity, not a line-for-line port — see GitHub issue #159). Depends only on the
/// `GoTrueTransport` protocol and `SecureSessionStore`, not on `URLSession` directly, so it stays
/// testable with a fake transport.
final class SupabaseAuthRepository: AuthRepository {
    private let transport: GoTrueTransport
    private let sessionStore: SecureSessionStore
    private let nowEpochSeconds: () -> Int64

    init(
        transport: GoTrueTransport,
        sessionStore: SecureSessionStore,
        nowEpochSeconds: @escaping () -> Int64 = { Int64(Date().timeIntervalSince1970) }
    ) {
        self.transport = transport
        self.sessionStore = sessionStore
        self.nowEpochSeconds = nowEpochSeconds
    }

    func requestEmailOtp(email: String) async -> AuthOutcome {
        switch await transport.requestEmailOtp(email: email) {
        case .ok, .sessionTokens:
            return .otpSent
        case .httpError(let status):
            return .retryableError(code: retryableCode(forStatus: status))
        case .networkError:
            return .retryableError(code: .network)
        case .malformed:
            return .retryableError(code: .unknown)
        }
    }

    func verifyEmailOtp(email: String, code: String) async -> AuthOutcome {
        switch await transport.verifyEmailOtp(email: email, code: code) {
        case .sessionTokens(let userID, let accessToken, let refreshToken, let expiresIn):
            return .success(session: buildSession(userID: userID, accessToken: accessToken, refreshToken: refreshToken, expiresInSeconds: expiresIn))
        case .httpError(let status) where (400...403).contains(status):
            return .invalidCode
        case .httpError(let status):
            return .retryableError(code: retryableCode(forStatus: status))
        case .networkError:
            return .retryableError(code: .network)
        case .ok, .malformed:
            return .retryableError(code: .unknown)
        }
    }

    func signInWithAppleIDToken(idToken: String, rawNonce: String) async -> AuthOutcome {
        switch await transport.signInWithIDToken(provider: "apple", idToken: idToken, nonce: rawNonce) {
        case .sessionTokens(let userID, let accessToken, let refreshToken, let expiresIn):
            return .success(session: buildSession(userID: userID, accessToken: accessToken, refreshToken: refreshToken, expiresInSeconds: expiresIn))
        case .httpError(let status) where (400...403).contains(status):
            return .unsupported
        case .httpError(let status):
            return .retryableError(code: retryableCode(forStatus: status))
        case .networkError:
            return .retryableError(code: .network)
        case .ok, .malformed:
            return .retryableError(code: .unknown)
        }
    }

    func refresh() async -> AuthOutcome {
        guard let stored = sessionStore.load() else {
            return .needsReauth
        }
        switch await transport.refresh(refreshToken: stored.refreshToken) {
        case .sessionTokens(let userID, let accessToken, let refreshToken, let expiresIn):
            return .success(session: buildSession(userID: userID, accessToken: accessToken, refreshToken: refreshToken, expiresInSeconds: expiresIn))
        case .networkError:
            return .retryableError(code: .network)
        case .httpError, .ok, .malformed:
            return .needsReauth
        }
    }

    func exchangeOAuthCode(providerId: String, code: String, codeVerifier: String) async -> AuthOutcome {
        switch await transport.exchangeOAuthCode(providerId: providerId, code: code, codeVerifier: codeVerifier) {
        case .sessionTokens(let userID, let accessToken, let refreshToken, let expiresIn):
            return .success(session: buildSession(userID: userID, accessToken: accessToken, refreshToken: refreshToken, expiresInSeconds: expiresIn))
        // 400/403 on OAuth exchange means the code is stale/reused (mismatch class); Supabase
        // distinguishes further only in the response body, which by M0c §8 the transport must
        // not expose. 404 means the provider isn't wired.
        case .httpError(let status) where (400...403).contains(status):
            return .oauthCallbackMismatch
        case .httpError(let status) where status == 404:
            return .oauthProviderNotConfigured
        case .httpError(let status):
            return .retryableError(code: retryableCode(forStatus: status))
        case .networkError:
            return .retryableError(code: .network)
        case .ok, .malformed:
            return .retryableError(code: .unknown)
        }
    }

    func startPasskeyAuthentication() async -> PasskeyStartResult<PasskeyAuthenticationOptions> {
        switch await transport.startPasskeyAuthentication() {
        case .options(let options):
            return .ready(options)
        case .httpError(let status):
            return .failed(code: retryableCode(forStatus: status))
        case .networkError:
            return .failed(code: .network)
        case .malformed:
            return .failed(code: .unknown)
        }
    }

    func signInWithPasskey(challengeID: String, assertion: PasskeyResult.Assertion) async -> AuthOutcome {
        switch await transport.verifyPasskeyAuthentication(challengeID: challengeID, assertion: assertion) {
        case .sessionTokens(let userID, let accessToken, let refreshToken, let expiresIn):
            return .success(session: buildSession(userID: userID, accessToken: accessToken, refreshToken: refreshToken, expiresInSeconds: expiresIn))
        case .httpError(let status):
            return .retryableError(code: retryableCode(forStatus: status))
        case .networkError:
            return .retryableError(code: .network)
        case .ok, .malformed:
            return .retryableError(code: .unknown)
        }
    }

    func startPasskeyRegistration() async -> PasskeyStartResult<PasskeyRegistrationOptions> {
        guard let session = sessionStore.load() else { return .needsReauth }
        switch await transport.startPasskeyRegistration(accessToken: session.accessToken) {
        case .options(let options):
            return .ready(options)
        case .httpError(let status) where status == 401 || status == 403:
            return .needsReauth
        case .httpError(let status):
            return .failed(code: retryableCode(forStatus: status))
        case .networkError:
            return .failed(code: .network)
        case .malformed:
            return .failed(code: .unknown)
        }
    }

    func registerPasskey(challengeID: String, credential: PasskeyRegistrationResult.Credential) async -> AuthOutcome {
        guard let session = sessionStore.load() else { return .needsReauth }
        switch await transport.verifyPasskeyRegistration(
            challengeID: challengeID,
            credential: credential,
            accessToken: session.accessToken
        ) {
        // Registration succeeds with 2xx-no-body; the app already has a session, so surface the
        // existing stored one (parity with Android SupabaseAuthRepository).
        case .ok:
            return .success(session: session)
        case .sessionTokens(let userID, let accessToken, let refreshToken, let expiresIn):
            return .success(session: buildSession(userID: userID, accessToken: accessToken, refreshToken: refreshToken, expiresInSeconds: expiresIn))
        case .httpError(let status) where status == 401 || status == 403:
            return .needsReauth
        case .httpError(let status):
            return .retryableError(code: retryableCode(forStatus: status))
        case .networkError:
            return .retryableError(code: .network)
        case .malformed:
            return .retryableError(code: .unknown)
        }
    }

    func listPasskeys() async -> PasskeyListOutcome {
        guard let session = sessionStore.load() else { return .needsReauth }
        switch await transport.listPasskeys(accessToken: session.accessToken) {
        case .passkeys(let passkeys):
            return .success(passkeys)
        case .httpError(let status) where status == 401 || status == 403:
            return .needsReauth
        case .httpError(let status):
            return .failed(code: retryableCode(forStatus: status))
        case .networkError:
            return .failed(code: .network)
        case .malformed:
            return .failed(code: .unknown)
        }
    }

    func deletePasskey(id: String) async -> PasskeyManagementOutcome {
        guard let session = sessionStore.load() else { return .needsReauth }
        switch await transport.deletePasskey(id: id, accessToken: session.accessToken) {
        case .ok:
            return .success
        case .httpError(let status) where status == 401 || status == 403:
            return .needsReauth
        case .httpError(let status):
            return .failed(code: retryableCode(forStatus: status))
        case .networkError:
            return .failed(code: .network)
        case .malformed, .sessionTokens:
            return .failed(code: .unknown)
        }
    }

    func signOut() async {
        sessionStore.clear()
    }

    private func buildSession(userID: String, accessToken: String, refreshToken: String, expiresInSeconds: Int) -> Session {
        Session(
            userID: userID,
            accessToken: accessToken,
            refreshToken: refreshToken,
            expiresAtEpochSeconds: nowEpochSeconds() + Int64(expiresInSeconds)
        )
    }
}

private func retryableCode(forStatus status: Int) -> RetryableCode {
    if status == 408 || status == 504 { return .timeout }
    if (500...599).contains(status) { return .server }
    return .unknown
}
